src/pow.rs

```diff
@@ -270,3 +323,33 @@ fn count_leading_zeros(data: &[u8]) -> u8 {
 
     leading_zeros
 }
+
+/// The faucet will dynamically adjust the difficulty based on the
+/// current balance of the faucet to make it increasingly difficult
+/// to retrieve funds from the faucet. The actual equation for this
+/// is:
+/// ```math
+/// \text{difficulty} = \left(\text{max\_difficulty} - \text{min\_difficulty}\right) \cdot \left(1 - \log_{\text{btc\_per\_emission}}\left(\frac{\text{balance} - \text{min\_balance}}{\text{min\_balance}}\right)\right)+\text{min\_difficulty}
+/// ```
+pub fn calculate_difficulty(
+    max_difficulty: f32,
+    min_difficulty: f32,
+    balance: Amount,
+    min_balance: Amount,
+    per_emission: Amount,
+) -> f32 {
+    // Ensure balance is at least min_balance to avoid invalid logarithm
+    if balance <= min_balance {
+        return max_difficulty; // Return max difficulty if balance is at or below min_balance
+    }
+
+    let balance = balance.to_btc() as f32;
+    let min_balance = min_balance.to_btc() as f32;
+    let per_emission = per_emission.to_btc() as f32;
+
+    let log_term = ((balance - min_balance) / min_balance).log(per_emission);
```

Comment:
Best to use checked division or otherwise test against `balance` vanishing. Otherwise you're relying on the interaction between undefined behaviors.

You could probably replace the subtraction with a checked or saturating variant instead of the initial sanity check. It'd be more annoying to write out, but I think it's generally safer.

Reply:
Oh weird; didn't know there weren't checked variants :(